### PR TITLE
More version-friendly Operation<,>.

### PR DESCRIPTION
**Pull request by jskeet**
If the API looks right, I'll add tests - so the review at the moment is "does this feel right".

Also raised https://github.com/google/protobuf/issues/3294 to improve Google.Protobuf.

### CODE REVIEW

File: apis/Google.LongRunning/Google.LongRunning/Operation.cs [hunk 1]
```diff
@@ -105,17 +129,37 @@ public Operation(Operation rpcMessage, OperationsClient client)
         /// Retrieves the metadata associated with this operation, or <c>null</c> if there is no
         /// metadata in the underlying response message.
         /// </summary>
-        public TMetadata Metadata => RpcMessage.Metadata?.Unpack<TMetadata>();
+        /// <remarks>
+        /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// See the <see cref="Operation"/> documentation for information about dealing with different metadata type versions.
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">Metadata is present, but is not of the expected type.</exception>
```
Comment (chrisdunelm) [inline apis/Google.LongRunning/Google.LongRunning/Operation.cs]: Use a custom exception `IncompatibleMetadataException` (or something), for clarity?

File: apis/Google.LongRunning/Google.LongRunning/Operation.cs [hunk 2]
```diff
@@ -105,17 +129,37 @@ public Operation(Operation rpcMessage, OperationsClient client)
...
+        /// If the operation has completed but the result is not present (for example due to being excluded by
+        /// a field mask) this returns null.
+        /// </summary>
+        /// <remarks>
+        /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// See the <see cref="Operation"/> documentation for information about dealing with different response type versions.
         /// </remarks>
         /// <exception cref="OperationFailedException">The operation completed with an error.</exception>
-        /// <exception cref="InvalidOperationException">The operation has not completed yet.</exception>
+        /// <exception cref="InvalidOperationException">The operation has not completed yet, or the result is present but
```
Comment (chrisdunelm) [inline apis/Google.LongRunning/Google.LongRunning/Operation.cs]: Return different Exceptions depending on which condition failed? (Related to above comment)

Comment (jskeet) [inline apis/Google.LongRunning/Google.LongRunning/Operation.cs]: It feels a bit overkilly to me - especially as it's easy enough to check whether or not it *has* completed, if you need to. (You could even check it in a guard clause if you really wanted.)

InvalidOperationException really does feel like the most appropriate one...

**Comment by jskeet** [comment]
Tests are now present, but I had to change the implementation as well. If the result has been masked out, we don't really know whether it completed successfully or whether it was an error that was masked out. So having the Done property *is* a useful thing.

@chrisdunelm if you could review today or Monday that would be great (before I lose context) but I may wait for internal discussions to conclude before merging.

File: apis/Google.LongRunning/Google.LongRunning/Operation.cs [hunk 3]
```diff
@@ -105,34 +126,74 @@ public Operation(Operation rpcMessage, OperationsClient client)
...
             get
             {
-                switch (RpcMessage.ResultCase)
+                // This uses the Done property...
+                if (!IsCompleted)
+                {
+                    throw new InvalidOperationException("Operation has not completed.");
+                }
+
+                // If we have no result case, that's could be because there was an error but the fieldmask
```
Comment (chrisdunelm) [inline apis/Google.LongRunning/Google.LongRunning/Operation.cs]: "that", not "that's"

Comment (jskeet) [inline apis/Google.LongRunning/Google.LongRunning/Operation.cs]: Doh, thanks. Will fix in a minute.

**Comment by jskeet** [comment]
Discussion concluded this is appropriate - we may make Any support even better in protobuf, which could have knock-on effects here later (e.g. fetch `object GetResult(TypeRegistry)`).

**Pull request merged**